tutorials/environmental_data_and_visualization.md

```diff
@@ -0,0 +1,173 @@
+\page environmental_data Visualizing Environment Data
+
+Lot of times you may come accross environment properties that vary with time and
+space. For instance if you'd like to model properties such as gas distribution
+and water temperature one needs to have some way of defining values such that
+the values vary through out the environment. Gazebo's Environment Data
+infrastructure enables such visualization and sensing capabilities.
+
+# Loading Environment Data
+
+Currently gazebo is able to load data from CSV files and display them. It is
+assumed the points are sampled at a fixed grid interval. For gaps in between the
+sample points, Gazebo will interpolate the data.
+
+## Data Format
+
+As mentioned earlier, the data should be a CSV file with grid based spacing.
+The first line of the CSV file should be the heading for each column.
+There should be a column for time (in seconds), the x coordinate, y coordinate
+and z coordinate (in meters). You may have multiple columns of data like so:
+```
+timestamp,humidity,x,y,z
+0,0,-1,-1,-1
+0,0,-1,-1, 1
+0,0,-1, 1,-1
+0,0,-1, 1, 1
+0,0, 1,-1,-1
+0,0, 1,-1, 1
+0,0, 1, 1,-1
+0,0, 1, 1, 1
+1,1,-1,-1,-1
+1,1,-1,-1, 1
+1,1,-1, 1,-1
+1,1,-1, 1, 1
+1,1, 1,-1,-1
+1,1, 1,-1, 1
+1,1, 1, 1,-1
+1,1, 1, 1, 1
+
+```
+For the x and y coordinates you may also use latitude and longitude in degrees.
+Heading names don't matter. Order also doesn't matter.
+```
+timestamp,humidity,latitude,longitude,altitude
+1658923062,91,36.80029505,-121.788972517,0.8
+1658923062,88,36.80129505,-121.788972517,0.8
+1658923062,89,36.80029505,-121.789972517,0.8
+1658923062,92,36.80129505,-121.789972517,0.8
+1658923063,90,36.80029505,-121.788972517,0.8
+1658923063,85,36.80129505,-121.788972517,0.8
+1658923063,87,36.80029505,-121.789972517,0.8
+1658923063,94,36.80129505,-121.789972517,0.8
+```
+> :warning: Note that only one Environment Data file per world is allowed. Also,
+Gazebo can only load data that fits into memory.
+
+## Pre-Loading
+If you are writing SDFormat files by hand or via a script you may want to
+pre-load the data. You may use the `EnvironmentPreload` plugin for this. For
+instance we could load the first file like so (by adding the environment to the
+):
+```xml
+    <plugin
+      filename="gz-sim-environment-preload-system"
+      name="gz::sim::systems::EnvironmentPreload">
+      <data>environmental_data.csv</data>
```

Comment:
@arjo129 nit: it may be worth mentioning how path resolution works here.

Reply:
+1
When I run
```
gz sim environmental_sensor.sdf
```
I get
```
[Err] [EnvironmentalSensorSystem.cc:460] No sensor data loaded
[Err] [EnvironmentPreload.cc:258] No environmental data file was found at environmental_data.csv
```
The csv file probably needs to be installed somewhere, to build/ or install/. I don't know which one `gz sim` is picking up, as there are three:
```
$ find . -name environmental_sensor.sdf
./src/gz-sim/examples/worlds/environmental_sensor.sdf
./install/share/gz/gz-sim7/worlds/environmental_sensor.sdf
./build/gz-sim7/test/worlds/environmental_sensor.sdf
```
But there are only 2 CSV files, only in src/:
```
$ find . -name environmental_data.csv
./src/gz-sim/test/worlds/environmental_data.csv
./src/gz-sim/examples/worlds/environmental_data.csv
```
gz sim is probably not finding them.

Even if I copy one of the CSV files to the current working directory, now I don't get the 2nd error about file not found, but I still get the first error "No sensor data loaded."